Add NativeFunctionGroup

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45990 Refactor dispatcher and native to use Signature structure.
* #45975 Remove unnecessary byte-for-byte compatibility code that is not needed.
* #45974 Rename legacy_dispatcher to native.
* #45973 Reorder dispatcher/legacy_dispatcher types
* **#45918 Add NativeFunctionGroup**
* #45890 Rewrite implementation of faithful cpp signatures

This groups together related native functions (functional, inplace, out)
into a single group.  It's not used by anything but Jiakai said this
would be useful for his stuff so I'm putting it in immediately.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24163526](https://our.internmc.facebook.com/intern/diff/D24163526)

ezyang committed:
Add NativeFunctionGroup

This groups together related native functions (functional, inplace, out)
into a single group.  It's not used by anything but Jiakai said this
would be useful for his stuff so I'm putting it in immediately.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add NativeFunctionGroup"


This groups together related native functions (functional, inplace, out)
into a single group.  It's not used by anything but Jiakai said this
would be useful for his stuff so I'm putting it in immediately.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

smessmer commented on tools/codegen/model.py:
```diff
@@ -199,6 +199,54 @@ def __post_init__(self) -> None:
 
+# Represents a bundle of native functions that are semantically related.
+@dataclass(frozen=True)
+class NativeFunctionGroup:
```
A third concept we also name "group"?

ezyang replied:
Second, really; after I got rid of `grouped_arguments` for packs in faithful C++ signature we are now just "group arguments" and "group native functions." IDK if you have a better name; in this case it seems less bad because the entity being grouped is different between the two cases.

ezyang committed:
Update on "Add NativeFunctionGroup"


This groups together related native functions (functional, inplace, out)
into a single group.  It's not used by anything but Jiakai said this
would be useful for his stuff so I'm putting it in immediately.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24163526](https://our.internmc.facebook.com/intern/diff/D24163526)

[ghstack-poisoned]

ezyang committed:
Update on "Add NativeFunctionGroup"


This groups together related native functions (functional, inplace, out)
into a single group.  It's not used by anything but Jiakai said this
would be useful for his stuff so I'm putting it in immediately.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24163526](https://our.internmc.facebook.com/intern/diff/D24163526)

[ghstack-poisoned]

ezyang committed:
Update on "Add NativeFunctionGroup"


This groups together related native functions (functional, inplace, out)
into a single group.  It's not used by anything but Jiakai said this
would be useful for his stuff so I'm putting it in immediately.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24163526](https://our.internmc.facebook.com/intern/diff/D24163526)

[ghstack-poisoned]
